### PR TITLE
Density matrix implementation

### PR DESCRIPTION
**Pull request by smorawetz**

### CODE REVIEW

File: qucumber/utils/cplx.py [hunk 1]
```diff
@@ -172,6 +172,31 @@ def outer_prod(x, y):
...
+    """
+
+    if len(list(x.size())) < 2:
+        raise ValueError("The input is not of the right dimension!")
+
+    if len(real(x)) == 0:
+        raise ("x must be non-empty!")
+
+    for i in range(len(real(x)) - 1):
+        x[:, i + 1] = scalar_mult(x[:, i], x[:, i + 1])
```
[the remark this answers is not in the document]
Comment (smorawetz) [reply qucumber/utils/cplx.py]: Good thinking, I hadn't caught on to this. However, this is actually something we could do away with completely, since I forgot to remove it after I stopped calling it in the main code. However, I can fix it and keep it in, if there is some reason to?

File: qucumber/rbm/purification_rbm.py [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+        """
+        if len(v.shape) < 2 and len(a.shape) < 2:
+            v = v.unsqueeze(0)
+            a = a.unsqueeze(0)
+
+            energy = torch.zeros(v.shape[0], a.shape[0], dtype=torch.double)
+
+            vb_term = torch.mv(v, self.visible_bias)
+            ab_term = torch.mv(a, self.aux_bias)
+            vb_term += torch.mv(torch.matmul(v, self.weights_U.data.t()), self.aux_bias)
```
[the remark this answers is not in the document]
Comment (smorawetz) [reply qucumber/rbm/purification_rbm.py]: Excellent catch, I've changed that. Although I don't think this is called in training though so it shouldn't affect that.

[the remark this answers is not in the document]
Comment (emerali) [reply qucumber/rbm/purification_rbm.py]: i believe it would affect the computed value of the partition function which would mean the fidelity calculations and other exact metrics are off by a little bit. This might also explain why the gradient testing is a bit unpredictable.

[the remark this answers is not in the document]
Comment (smorawetz) [reply qucumber/rbm/purification_rbm.py]: Partition function computed by trace of constructed matrix. Still this change has been made if the partition function is every required

File: qucumber/utils/cplx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (smorawetz) [reply qucumber/utils/cplx.py]: Removed from code as it is not used anywhere and would only cause confusion by its presence.

**Comment by emerali** [comment]
Note for future reference: merging despite CI fails due to the fact that develop branch is currently out of date and so is its CI config.
